packages/zoe/src/contractFacet/contractFacet.js

```diff
@@ -350,6 +350,28 @@ export function buildRootObject() {
       getBrandForIssuer: issuer => issuerTable.getByIssuer(issuer).brand,
       getIssuerForBrand: brand => issuerTable.getByBrand(brand).issuer,
       getAmountMath,
+      /**
+       * Provide a jig object for testing purposes only.
+       *
+       * The contract code provides a callback whose return result will
+       * be made available to the test that started this contract. The
+       * supplied callback will only be called in a testing context,
+       * never in production; i.e., it is only called if `testJigSetter`
+       * was supplied.
+       *
+       * If no, \testFn\ is supplied, then an empty jig will be used.
+       * An additional `zcf` property set to the current ContractFacet
+       * will be appended to the returned jig object (overriding any
+       * provided by the `testFn`).
+       *
+       * @type SetTestJig
+       */
+      setTestJig: (testFn = () => ({})) => {
+        if (testJigSetter) {
+          console.warn('TEST ONLY: capturing test data', testFn);
+          testJigSetter({ ...testFn(), zcf });
+        }
```

Comment:
I think this is a bit more complex than necessary, but maybe there are just use cases that aren't exercised by the tests added in this particular PR.

I'm thinking that `setTestJig()` should accept an object as its argument, and call `testJigSetter({...obj, zcf})` rather than taking a function. When I see functions passed into a method like this, I assume that either the caller wants to get something back (but `testFn()` takes no arguments), or they need to know *when* something has happened (but here, it always happens right away), or they need to know *whether* something has happened or not (exposed here, but maybe shouldn't be), or they have a lot of work to do and want to avoid it unless someone really cares (which I don't think is the case).

It might be nice if `setTestJig` were only exposed on the return value if `testJigSetter` was provided, but maybe your intention is to prevent contracts from being able to tell whether it was provided or not (by always providing it, but calling it has no effect unless `testJigSetter` was there).

It might be nice to call `setTestJig(zcf)` immediately, rather than relying upon some test contract to call it later. But obviously you plan to let contracts provide other data too. What about having the third `buildRootObject` be `testJig`, have this section call `testJig.setZCF(zcf)` right away (maybe `testJig?.setZCF(zcf)` is the modern/future way to guard against `testJig===undefined`?), and then wrap and share a `testJig.setContractData()` into the exported object for contracts to (optionally) share additional details if/when they care to?

Reply:
The reason for taking a function rather than an argument is so that production code doesn't do any work of setting up for testing if it's in production mode. While that will often just be passing in some variables, it might involve more effort that would be inappropriate in production context. With the current approach, you could imagine the babel transform stripping `setTestJig` invocations.

The reason we don't want to play games with whether `setTestJig` is exposed is that would just require there to be more test-only code in the production contract to see whether it's supposed to be setting up test info.

By having the invocation explicit and not in the return value, more things could be added to the jig over time (e.g., each offer might add something), it can be stripped automatically without changing the visible interface of the contract, and it doesn't go into the replay logs or other persistent state in production. Test jigs are internal details of the module.